### ISSUE TITLE
new rule: React Enforce Immutable State

### ISSUE THREAD

**Issue by HamletDRC** [microsoft/tslint-microsoft-contrib#38]
new rule: react-enforce-immutable-state

This rule should trigger violations when it appears that the user is mutating state within a React component. 

Examples: 

```
class MyComponent {
    private someMethod() {
        this.state.x = 'value'; // triggers violation
        this.state.x.y.z[0] = 'value'; // triggers violation
        this.state.model.setValue('value'); // triggers violation
        this.state.x.y.z.setValue('value'); // triggers violation
    }
}
```

React state should be immutable. It can subtly (and not so subtly) break React rendering when state is modified without using setState()/replaceState()

**Comment by loicraux** [comment]
Should also forbid `this.state = xxx` statements.

But all forbidden statements listed above should be allowed in constructor of es6 classes for React components.

**Comment by HamletDRC** [comment]
Thanks @loicraux .

**Comment by JoshuaKGoldberg** [comment]
### 💀 _**It's time!**_ 💀

**TSLint is being deprecated**; therefore, it and tslint-microsoft-contrib are no longer accepting pull requests for major new changes or features. See #4534. 😱

If you'd like to see this change implemented, you have two choices:
* **Recommended:** Check if this is available in [ESLint](https://eslint.org) + [typescript-eslint](https://typescript-eslint.io) ✅
* _Not Recommended: Fork TSLint or tslint-microsoft-contrib locally_ 🤷‍♂️

👋 It was a pleasure open sourcing with you!

_If you believe this message was posted here in error, please comment so we can re-open the issue!_
